colflow: propagate concurrency info from vectorized to FlowBase

yuzefovich opened the pull request:
**colflow: propagate concurrency info from vectorized to FlowBase**

We've recently merged a change to introduce concurrency in the local
flows. Those new concurrent goroutines are started by the vectorized
parallel unordered synchronizer, and `FlowBase` isn't aware of them; as
a result, `FlowBase.Wait` currently might not wait for all goroutines to
exit (which is an optimization when there are no concurrent goroutines).
This commit fixes the problem by propagating the information from the
vectorized flow to the FlowBase.

Addresses: #69419.

Release note: None (no stable release with this bug)

Release justification: bug fix to new functionality.

**sql: loosen up the physical planning of parallel scans**

This commit makes it so that in case we try to plan parallel
TableReaders but encounter an error during planning, we fallback to
having a single TableReader.

Release note: None

Release justification: bug fix to new functionality.

yuzefovich committed:
colflow: propagate concurrency info from vectorized to FlowBase

We've recently merged a change to introduce concurrency in the local
flows. Those new concurrent goroutines are started by the vectorized
parallel unordered synchronizer, and `FlowBase` isn't aware of them; as
a result, `FlowBase.Wait` currently might not wait for all goroutines to
exit (which is an optimization when there are no concurrent goroutines).
This commit fixes the problem by propagating the information from the
vectorized flow to the FlowBase.

Release note: None (no stable release with this bug)

Release justification: bug fix to new functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69612)
<!-- Reviewable:end -->

yuzefovich committed:
sql: loosen up the physical planning of parallel scans

This commit makes it so that in case we try to plan parallel
TableReaders but encounter an error during planning, we fallback to
having a single TableReader.

Release note: None

Release justification: bug fix to new functionality.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69612#-:-MiRk0EP7qMkGs6gRBrG:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/flowinfra/flow.go, line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69612#-MiRiApl1LyLTpmuDHYA:-MiRiApl1LyLTpmuDHYB:b7egptp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6774920ee34f2e8306a0837011130fa2ad3c4ab/pkg/sql/flowinfra/flow.go#L223)):*
> ```Go
> }
> 
> // SetStartedGoroutines updates the FlowBase about the fact whether any
> ```

nit: "about the fact whether" could be rephrased 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69612#-:-MiRrJIC0Bc01VNiEpoQ:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)



<!-- Sent from Reviewable.io -->
